Make the insert tags in picker providers configurable

Toflar opened the pull request:
I just had the problem that I wanted to use the page picker but not with the default `{{link_url::<value>}}` insert tag because I need absolute links, so I want the `|absolute` flag to always be added.
I noticed that the insert tag is unfortunately hard coded which I find a very limiting factor for a general picker provider.
So here we go, you can now specify whatever insert tag you like by specifying 

```php
'eval' => [
    'dcaPicker' => [
        'insertTag' => '{{my_super_insert_tag::%s}}',
    ],
],
```

which will take the insert tag for **all** picker providers (news, page etc.), though.
However, you can also do this:

```php
'eval' => [
    'dcaPicker' => [
        'pagePicker' => [
            'insertTag' => '{{link_url::%s}}',
        ],
        'newsPicker' => [
            'insertTag' => '{{news_url::%s}}',
        ],
    ],
],
```

I've implemented it for **all** the core pickers so it will also work for news, faq, events etc. not just for the page picker provider 😊 
Also there's a new `$config->getExtraForProvider()` so pickers can implement fetching provider specific extras falling back to general extras if they want to.

Toflar committed:
Make the insert tags in picker providers configurable

Toflar commented:
I just noticed an issue here. I need to be able to specify the extras to be only applicable for e.g. the page picker but not the news picker.
How do I do that? @contao/developers

Toflar committed:
Allow to configure extras only for specific providers

Toflar commented:
Updated PR to support extras per provider and also updated the issue description to demonstrate the usage 😄

leofeyer commented on calendar-bundle/src/Picker/EventPickerProvider.php:
```diff
@@ -45,7 +47,9 @@ public function supportsContext($context): bool
     public function supportsValue(PickerConfig $config): bool
     {
-        return false !== strpos($config->getValue(), '{{event_url::');
+        $insertTagChunks = explode('%s', $this->getInsertTag($config, self::DEFAULT_INSERTTAG), 2);
```
Can we encapsulate this in a `getInsertTagChunk()` method so we do not have to use multiple `explode()` calls?

Toflar replied:
I'm not even sure if that should be named `insertTag` or if that should be something else? I mean, in fact it's just whatever string you want to provide and we're going to replace '%s' with the value. I just don't know if there's any other use case other than an insert tag 😄 wdyt?

leofeyer replied:
> I just don't know if there's any other use case other than an insert tag

I cannot think of one.

Toflar replied:
Done 😄

Toflar committed:
Added an AbstractPickerProvider::getInsertTagChunks() to prevent repeated code

aschempp commented:
Instead of splitting the default inserttag by `%s`, why not simply store the prefix + suffix in the class constant? Or even more so have a `_PREFIX` and `_SUFFIX`?

Toflar commented:
Because it makes no sense to me. `%s` as the placeholder for `sprintf()` is well known so I don't really want to configure a value prefix and suffix in the dca configuration.

aschempp commented:
> I don't really want to configure a value prefix and suffix in the dca configuration.

Yeah, I would configure an array with two values for that case. I don't see why we would need to split on every use instead of already having an array.

leofeyer commented on core-bundle/src/Picker/AbstractPickerProvider.php:
```diff
@@ -131,6 +131,15 @@ protected function getInsertTag(PickerConfig $config, string $default): string
+     * Provides a shortcut get the "insertTag" extra value for child classes and
+     * split them at the placeholder (%s).
+     */
+    protected function getInsertTagChunks(PickerConfig $config, string $default): array
```
Can we not set a default value for the second argument?

```php
string $default = static::DEFAULT_INSERTTAG
```

Due to the late static binding, this should be set to the value of the constant of the child class.

Toflar replied:
And what if I forget to set this in my child? I would only notice it at runtime which makes the whole idea of an `abstract` class and interfaces pointless 😄

leofeyer replied:
It would use the constant of the parent class then.

Toflar replied:
Which does not exist because what value would it provide? 😄

leofeyer replied:
Ok, then why did you make the constant protected instead of private? Also, the naming is misleading: `DEFAULT_INSERTTAG` implies that there are constants for other, non-default insert tags, which there are not. We should change it to just `INSERTTAG`.

Toflar replied:
So that they can be replaced by a class extending a specific picker? I don't think the name is misleading, it's the default insert tag if no other configuration was provided.

leofeyer commented:
> I just noticed an issue here. I need to be able to specify the extras to be only applicable for e.g. the page picker but not the news picker.

Is this still an unsolved problem?

Toflar commented:
> Is this still an unsolved problem?

<img width="834" alt="Bildschirmfoto 2019-06-05 um 13 45 37" src="https://user-images.githubusercontent.com/481937/58954006-3aa23b00-8798-11e9-85a1-b11850e7be66.png">

leofeyer reviewed (changes requested):
Besides what I have already commented, there are still some major issues with the implementation that I do not like.

1. The whole "insert tag chunks" logic is quite cumbersome IMHO. It turns a readable `'{{event_url::'` into `$this->getInsertTagChunks($config, self::DEFAULT_INSERTTAG)[0]`.

2. I find it completely wrong that we have to pass the insert tag constant in the `getInsertTag()` method just so the abstract parent class knows about the insert tag. If the parent class could access the constant (or call a method) of the child class, we could simplify `getInsertTag($config, self::$DEFAULT_INSERTTAG)` to `getInsertTag($config)`. The same applies to the `getInsertTagChunks()` method.

3. The constant `DEFAULT_INSERTTAG` is misleading, because it suggests that there are other, non-default constants. I suggest renaming it to `INSERTTAG`.

Toflar committed:
Reworked concept a little bit

Toflar commented:
> 1. he whole "insert tag chunks" logic is quite cumbersome IMHO. It turns a readable `'{{event_url::'` into `$this->getInsertTagChunks($config, self::DEFAULT_INSERTTAG)[0]`.

Yeah well, it's the whole point of this PR though. We want to make something that is hardcoded dynamic so I'm sorry but I cannot really change this fact.

> 2. I find it completely wrong that we have to pass the insert tag constant in the getInsertTag() method just so the abstract parent class knows about the insert tag. If the parent class could access the constant (or call a method) of the child class, we could simplify getInsertTag($config, self::$DEFAULT_INSERTTAG) to getInsertTag($config). The same applies to the getInsertTagChunks() method.

I agree with that. I've reworked the concept in 418c264cb1d5762a8278d7261f1d2ce56b77d7f6. Do you like that better now? Of course, I cannot make the `getFallbackInsertTag()` method `abstract` because that would mean every picker that extends the `AbstractPickerProvider` would need to implement. This would be a) a BC break and b) it's not necessary anyway as you only have to provide it if you want to work with insert tags.

> 3. The constant DEFAULT_INSERTTAG is misleading, because it suggests that there are other, non-default constants. I suggest renaming it to INSERTTAG.

I've gotten rid of the constants completely and I also moved to the term "fallback insert tag" rather than "default insert tag".

leofeyer commented:
What was wrong with the constants?

```php
class AbstractPickerProvider
{
    public const INSERTTAG = null;

    public function getInsertTag(): string
    {
        if (null === static::INSERTTAG) {
            throw new \LogicException('Please add a public INSERTTAG constant in your picker provider class');
        }
        
        return static::INSERTTAG;
    }
}

class PagePickerProvider extends A
{
    public const INSERTTAG = '{{page::%s}}';
}
```

See https://3v4l.org/pRVpe

Toflar committed:
Bring back the constants

Toflar commented:
Adjusted accoringly. I made them protected though because they don't need to be accessed from the outside.

leofeyer committed:
Fix the coding style

leofeyer committed:
Add a getValue() method that returns the value without the surrounding insert tag chunks

leofeyer committed:
Add the isMatchingTag() method

Toflar commented:
@leofeyer thinking of it, maybe we should have an `AbstractInsertTagsPickerProvider` that extends the `AbstractPickerProvider` and provides the additional methods. Imho that would be cleaner design because then you don't get any insert tag related helper methods in autocompletion when you don't want to work with insert tags in your picker provider.

leofeyer commented:
Is there a picker that does not use insert tags?

Toflar commented:
Sure! Have a look at https://github.com/terminal42/contao-bynder/blob/master/src/Picker/BynderAssetPickerProvider.php for example :) This one is not even extending the AbstractPickerProvider :D

leofeyer committed:
Ensure that the insert tag related methods have "InsertTag" in their name

leofeyer committed:
Encapsulate the insert tag related methods in a separate class

leofeyer commented:
Implemented in b82b5add980b8fe1a5906f7833e30bc988fbddce.

Toflar commented:
I like that better, wdyt :)

leofeyer commented:
I like it, too. Gonna run one last test and then I would merge it if you want.

Toflar commented:
Go for it 👍 Thank you!

leofeyer commented:
Thank you @Toflar.

aschempp commented:
Sorry, but this is a BC break as well as a bad idea. Removing protected methods from a class that is **supposed** to be extended is not semver. Also, we should never expect a class constant, that's what abstract methods are for!

Toflar commented:
Nobody removed any methods?

aschempp commented:
wtf, looks like I had an outdated view of the PR. The constant still remains though. Can we use an abstract method instead of the constant please? This way all IDEs will tell you to implement that method.

Toflar commented:
I agree that would be better :)

leofeyer commented:
I'm going to change it.

leofeyer commented:
Changed in 92f4a9e759d8c70fb99fcedff7888bc420aeb99f.
